oap-server/server-storage-plugin/storage-influxdb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/influxdb/query/AggregationQuery.java

```diff
@@ -0,0 +1,152 @@
+                                                   Downsampling downsampling,
+                                                   long startTB, long endTB, Order order) throws IOException {
+        return getTopNEntity(
+            downsampling, indName,
+            subQuery(ServiceInstanceInventory.SERVICE_ID, serviceId, indName, valueCName, startTB, endTB), order, topN
+        );
+    }
+
+    @Override
+    public List<TopNEntity> getAllEndpointTopN(String indName, String valueCName, int topN, Downsampling downsampling,
+                                               long startTB, long endTB, Order order) throws IOException {
+        return getTopNEntity(downsampling, indName, subQuery(indName, valueCName, startTB, endTB), order, topN);
+    }
+
+    @Override
+    public List<TopNEntity> getEndpointTopN(int serviceId, String indName, String valueCName, int topN,
+                                            Downsampling downsampling,
+                                            long startTB, long endTB, Order order) throws IOException {
+        return getTopNEntity(
+            downsampling, indName,
+            subQuery(EndpointInventory.SERVICE_ID, serviceId, indName, valueCName, startTB, endTB), order, topN
+        );
+    }
+
+    private List<TopNEntity> getTopNEntity(Downsampling downsampling,
+                                           String name,
+                                           SelectSubQueryImpl<SelectQueryImpl> subQuery,
+                                           Order order,
+                                           int topN) throws IOException {
+        String measurement = ModelName.build(downsampling, name);
+        Comparator<TopNEntity> comparator = DESCENDING;
+        String functionName = "top";
+        if (order == Order.ASC) {
+            functionName = "bottom";
+            comparator = ASCENDING;
+        }
+
+        SelectQueryImpl query = select().function(functionName, "mean", topN).as("value")
+                                        .column(MetricsDAO.TAG_ENTITY_ID)
+                                        .from(client.getDatabase(), measurement);
+        query.setSubQuery(subQuery);
+
+        List<QueryResult.Series> series = client.queryForSeries(query);
+        if (log.isDebugEnabled()) {
+            log.debug("SQL: {} result set: {}", query.getCommand(), series);
+        }
+        if (series == null || series.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        List<List<Object>> dataset = series.get(0).getValues();
+        List<TopNEntity> entities = Lists.newArrayListWithCapacity(dataset.size());
+        dataset.forEach(values -> {
+            final TopNEntity entity = new TopNEntity();
+            entity.setId((String) values.get(2));
+            entity.setValue(((Double) values.get(1)).longValue());
+            entities.add(entity);
+        });
+        Collections.sort(entities, comparator);
```

Comment:
As you mentioned last time, the result is unordered. Could you add the comments some places to ease the readers?